src/core/exec-invoke.c

```diff
@@ -2096,31 +2096,45 @@ static int setup_private_users(uid_t ouid, gid_t ogid, uid_t uid, gid_t gid) {
          * For unprivileged users (i.e. without capabilities), the root to root mapping is excluded. As such, it
          * does not need CAP_SETUID to write the single line mapping to itself. */
 
+        if (private_users == PRIVATE_USERS_IDENTITY) {
```

Comment:
If private_users is passed to the function anyway, I think it should also handle `_OFF`.

Reply:
It's not trivial to handle `_OFF` within the function. You can't just skip the function, it still needs to do its operation when running unprivileged (when it is invoked the first time).

Reply:
at the very least there should be an `assert(IN_SET(private_users, …))` at the top of the function

Reply:
@poettering But it can be all of them, that's the thing. In the unprivileged case, we call `setup_private_users()` with `private_users` set to `PRIVATE_USERS_OFF`

Reply:
oh, we do?

Reply:
oh, right, we do.

Reply:
hmm, maybe add a comment to the top of the function:

```c
/* Note that this function may even be invoked for private_users == PRIVATE_OFF, because the PrivateUsers= logic is implicitly enabled when various privileged (mostly mount-related) operations are requested on a unit invoked by an unprivileged service manager instance */
```
or so.

Reply:
How about this?